src/metabase/queries/models/card.clj

```diff
@@ -1238,33 +1240,61 @@
 
 ;;;; ------------------------------------------------- Search ----------------------------------------------------------
 
+(defn- dataset-query->dimensions
+  "Extract dimensions (non-aggregation columns) from a dataset query."
+  [dataset-query-str]
+  (when dataset-query-str
+    (lib.metadata.jvm/with-metadata-provider-cache
+      (let [dataset-query     (json/decode+kw dataset-query-str)
+            metadata-provider (lib.metadata.jvm/application-database-metadata-provider (:database dataset-query))
+            lib-query         (lib/query metadata-provider dataset-query)
+            columns           (lib/returned-columns lib-query)]
```

Comment:
I'll assume these APIs are OK to use here, but might be worth checking with someone more knowledgeable about metabase.lib if you haven't already